### DISCUSSION TITLE
How to get timestamps?

### DISCUSSION THREAD

**Discussion by voidpointertonull** [axboe/liburing#827] [Q&A]
While moving to using io_uring was a blessing for performance, I've started to face the problem that without using at least one thread waiting for event completion, I don't seem to be able to get good performance info since there's no timing information on when specific operations are done.
Went through the following steps:
- Looked into (ab)using timeouts as it might be ideal to use them anyway to avoid I/O operations getting stuck for too long, but apparently there's no cancellation timestamp
- Went through the opcode list to see if any of the operations places a timestamp in the result, but didn't really find one
- Figured that as there's an abundance of information represented with files, maybe there's one somewhere providing a timestamp, but couldn't find one that's surely intended to be used as such

While I can think of some perverted approaches like OPEN+STATX+CLOSE with a temporary file providing a timestamp when linked to the operation (chain), it's quite expensive and likely has portability issues. Am I missing something, or is there really no good way to essentially do profiling by getting timestamps without involving userland with synchronous waiting for the event completion that's desired to be timestamped?

**Comment by axboe** [comment]
I guess it depends on what your use case is - is it doing this all the time, or is it for specific testing/profiling of workloads? For the latter, we do have trace events that can trace many different things, completions being one of them.

**Comment by voidpointertonull** [comment]
My use case would be continuous monitoring of peak performance for general fitness of the system, most importantly serving as a way to identify a bottleneck preferably before it becomes a problem. It's mostly a need arising from 2 or more possible bottlenecks being present with a heavily io_uring driven setup, so event completions are generally not monitored within reasonable time but tended to only to maintain high throughput without having a low latency completion -> timestamping step.

Consider the example of a file server with an efficient and simple setup of one thread just using io_uring to do the heavy lifting where without performance insights, the following problems could arise without monitoring:
- Slow performance degradation wouldn't be caught until the degrading part becomes the new bottleneck. Storage devices could still perform well, but maybe the page cache hitrate just went down significantly with size increase, files got significantly fragmented, and so on.
- Sudden performance degradation requires a longer and/or deeper investigation to find the cause instead of performance info pointing quickly at least in one of the two likely directions, or maybe even both for a "host is on fire/cursed" kind of situation
- Erratic behavior like unreasonable latency spikes would not be caught until the device with the issue becomes the bottleneck, making for unpleasant surprises in the future which could have been avoided

Of course file servers don't tend to have such performance monitoring, but the examples are representative of real world issues. Most of my work is related to the HPC field where there's a constant desire of monitoring performance even of parts which aren't the bottleneck yet, especially as the bottleneck often moves around due to active development.

With the usual threaded approach, as different devices tend to have different threads and blocking syscalls, performance measurement and reporting is not a problem there, so let's say with a simple test setup running a more complex workload would happily report peak performances like:
- Network: 1100 MiB/s
- Disk: 5500 MiB/s
- Accelerator card: 8000 MiB/s

However if I move to an efficient io_uring setup where the ring is always just kept fed until the work is done, all I could reliably tell is that all 3 possible bottlenecks reach at least 1100 MiB/s with no idea which device is actually the bottleneck, having no information about which part should be scaled further without deeper inspection, and generally losing the ability to report the magic numbers most decisions are based on. An important consideration here is that the programs tend to be in flexible environment regularly changing both due to development and to suit changing needs, so there's no one time performance test and deployment but it's closer to continuous testing of various hardware setups.

Not sure if the explanation is too convincing as it may be considered somewhat niche performance worshipping as most programs showing performance info reveal only the total/output performance, then the rest is an adventure left for the user, and that seems to be considered good enough for most.
However the problem itself is quite simple as it's just a matter of some strategically placed timestamping generally done before and after the relevant syscall, so I was just left head scratching how to "port" this approach, but only came up with questionable workarounds, and the idea of just simply "downgrading" to the separate thread waiting for io_uring syscall exit to catch completion as soon as possible for reliable timestamp.

Generally it seems like that at least networking has a whole lot of timestamping goods, so the desire is definitely present, and the network specific solution is presenting the usual issue that file I/O interfaces are still lacking in comparison, although it would be probably better to have at least one simple generic solution for both. #444 might be relevant which also seems to have sought timestamping, but ended up with the network specific approach.

I really wondered if I just missed something as even with just a single bottleneck, calculating progress from elapsed time (based on timestamps) and processed data is the bread and butter of even progress bars showing either performance or inferring remaining time from that, but given large operations (and therefore low granularity) for efficiency, the assumption of the last completion seen being fresh leads to quite significant inaccuracy if there isn't one thread that really doesn't have to do much else than tending to the rings. Haven't considered if it's a solution fitting for everyone, but linking a timestamping operation to everything I want to measure seemed to be the straight forward way of getting timing info that's even more precise than getting a timestamp after syscall exit, and would have no overhead when not used, so I'm surprised that it's currently not possible which is why I was asking if I'm really just missing something, even if it's my approach being wrong for some reason.
